ocaml/xapi/xapi_host_driver.ml

```diff
@@ -0,0 +1,297 @@
+    let updated_drivers =
+      DriverSet.inter db_drivers_set filesystem_drivers
+      |> DriverSet.elements
+      |> List.filter_map (fun driver_name ->
+             let self = DriverMap.find driver_name db_drivers_map in
+             let selected_version, active_version, driver_versions =
+               get_driver_versions driver_name driver_paths.selected_dir
+                 driver_paths.all_versions_dir
+             in
+             let selected_changed =
+               Db.Host_driver.get_selected_version ~__context ~self
+               <> selected_version
+             in
+             let active_changed =
+               Db.Host_driver.get_active_version ~__context ~self
+               <> active_version
+             in
+             let versions_changed =
+               Db.Host_driver.get_versions ~__context ~self <> driver_versions
+             in
+             if selected_changed then
+               Db.Host_driver.set_selected_version ~__context ~self
+                 ~value:selected_version ;
+             if active_changed then
+               Db.Host_driver.set_active_version ~__context ~self
+                 ~value:active_version ;
+             if versions_changed then
+               Db.Host_driver.set_versions ~__context ~self
+                 ~value:driver_versions ;
+             if selected_changed || active_changed || versions_changed then
+               Some driver_name
+             else
+               None
+         )
+    in
+    (* Remove drivers that are no longer in the filesystem from the DB *)
+    let removed_drivers =
+      DriverSet.diff db_drivers_set filesystem_drivers
+      |> DriverSet.elements
+      |> List.map (fun driver_name ->
+             let self = DriverMap.find driver_name db_drivers_map in
+             destroy ~__context ~self ; driver_name
+         )
+    in
+    (* Add the new drivers *)
+    let added_drivers =
+      DriverSet.diff filesystem_drivers db_drivers_set
+      |> DriverSet.elements
+      |> List.map (fun driver_name ->
+             let selected_version, active_version, driver_versions =
+               get_driver_versions driver_name driver_paths.selected_dir
+                 driver_paths.all_versions_dir
+             in
+             create ~__context ~host ~name_label:driver_name
+               ~versions:driver_versions ~active_version ~selected_version ;
+             driver_name
+         )
+    in
+    [
+      ("updated_drivers", updated_drivers)
+    ; ("removed_drivers", removed_drivers)
+    ; ("added_drivers", added_drivers)
+    ]
+    |> List.filter (fun (_, ds) -> ds <> [])
+  with
+  | Sys_error e ->
+      D.debug "Failed to parse drivers on host %s: %s" (Ref.string_of host) e ;
+      []
+  | e ->
+      D.debug "Failed to parse drivers on host %s: %s - %s" (Ref.string_of host)
+        (Printexc.to_string e)
+        (Printexc.get_backtrace ()) ;
+      []
+
+(* Runs on the necessary host *)
+let select ~__context ~self ~version =
```

Comment:
Would the `select`/`deselect` need a protection from being invoked concurrently?

Reply:
Hmm. I think the symbol link under `/lib/modules/<kernel-version>/updates/` needs protection as well.

Reply:
This is somewhat harder than a single mutex per reload, since we'd have to protect a particular driver name - a different driver could still be changed concurrently, right? I do not think anything particularly harmful can come from concurrent removal/symlinking of the same driver - the last reload will pick up the last changes anyway.

Reply:
> the last reload will pick up the last changes anyway.

I see. Thanks.